crates/media-parser/src/format/mp3/metadata.rs

```diff
@@ -0,0 +1,275 @@
+//! # MP3 Metadata Extraction
+//!
+//! Extracts metadata from MP3 files using ID3v2 tags.
+//!
+//! ## ID3v2 Frame IDs
+//!
+//! - TIT2: Title
+//! - TPE1: Artist
+//! - TALB: Album
+//! - TYER/TDRC: Year
+//! - TRCK: Track number
+//! - TCON: Genre
+//! - COMM: Comment
+//! - TCOM: Composer
+
+use super::duration::calculate_duration;
+use super::tags::{frame_id_to_key, frame_name};
+use crate::Result;
+use crate::errors::MediaParserError;
+use crate::helpers::{
+   decode_latin1, decode_utf8, decode_utf16_be, decode_utf16_with_bom, trim_null_and_whitespace,
+};
+use crate::stream::StreamReader;
+use crate::types::{Meta, Metadata};
+
+const ID3_HEADER_SIZE: usize = 10;
+const ID3_FRAME_HEADER_SIZE: usize = 10;
+
+/// Reads metadata from an MP3 file.
+///
+/// Extracts ID3v2 tags and calculates duration.
+///
+/// # Errors
+///
+/// Returns an error if:
+/// - The file does not have a valid ID3v2 header
+/// - Reading from the stream fails
+pub async fn read_metadata(reader: &dyn StreamReader) -> Result<Metadata> {
+   let header = read_id3_header(reader).await?;
+   let frames = read_id3_frames(reader, &header).await?;
+   let values = frames_to_meta(frames);
+
+   // Calculate duration from audio frames
+   let id3_end = ID3_HEADER_SIZE as u64 + header.tag_size as u64;
+   let duration = calculate_duration(reader, id3_end).await?;
+
+   Ok(Metadata {
+      format: "MP3".to_string(),
+      values,
+      timescale: 1000,
+      duration: duration.millis,
+   })
+}
+
+#[derive(Debug)]
+struct Id3Header {
+   version: (u8, u8),
+   #[allow(dead_code)]
+   flags: u8,
+   tag_size: u32,
+}
+
+/// Reads and parses the ID3v2 header.
+async fn read_id3_header(reader: &dyn StreamReader) -> Result<Id3Header> {
+   let mut header_bytes = [0u8; ID3_HEADER_SIZE];
+   let bytes_read = reader.read_at(0, &mut header_bytes).await?;
+
+   if bytes_read < ID3_HEADER_SIZE {
+      return Err(MediaParserError::InvalidFormat(
+         "File too small for ID3 header".to_string(),
+      ));
+   }
+
+   if &header_bytes[0..3] != b"ID3" {
```

Comment:
This rejects valid MP3 files that don't have ID3v2 tags (they might have ID3v1 at the end, or no tags at all). The duration calculation should still work - consider making ID3 parsing optional.